docs/microbenchmark-design-guidelines.md

```diff
@@ -0,0 +1,557 @@
+* BenchmarkDotNet removes outliers by default (this repo is configured to remove only the upper outliers)
+* BenchmarkDotNet creates a type which derives from type with benchmarks. So the type with benchmarks must **not** be **sealed** and it can **not** be **static** and it has to be **public**. It also has to be a `class` (no structs support).
+
+## Setup
+
+Let's write a benchmark that measures the performance of reverting an array of 1 000 integers:
+
+```cs
+[Benchmark]
+public void Reverse()
+{
+    int[] array = Enumerable.Range(0, 1_000).ToArray();
+
+    Array.Reverse(array);
+}
+```
+
+Profile it using the [ETW Profiler](./benchmarkdotnet.md#Profiling):
+
+```cmd
+dotnet run -c Release -f netcoreapp3.0 --filter *.Reverse --profiler ETW
+```
+
+And open the produced trace file with [PerfView](https://github.com/Microsoft/perfview):
+
+![Regression](img/setup_array_reverse_profiler.png)
+
+As you can see, reversing the array took only `26.6%` of the benchmark execution time!! The rest was spent on executing the array creation logic. What does it mean? It means that the given benchmark is measuring the performance of creating and reversing the array. Not just reversing the array.
+
+This is why **the initialization logic should be always separated from the benchmark**. If possible of course.
+
+### GlobalSetup
+
+Public method marked with `[GlobalSetup]` attribute is going to be executed **exactly once**, before running the benchmark for the first time.
+
+```cs
+private int[] _array;
+
+[GlobalSetup]
+public void SetupReverse() => _array = Enumerable.Range(0, 1000).ToArray();
+
+[Benchmark]
+public void Reverse() => Array.Reverse(_array);
+```
+
+If given `[GlobalSetup]` does not specify the `Target(s)` benchmark name(s), it's executed for every benchmark declared in a given class.
+
+In case you want to have a class with multiple benchmarks and multiple, dedicated setup methods you need to use `Target` or `Targets` properties of the setup attribute:
+
+```cs
+[GlobalSetup(Target = nameof(Reverse))]
+public void SetupReverse()
+
+[GlobalSetup(Targets = new [] { nameof(Array), nameof(Span), nameof(ReadOnlySpan)})]
+public void SetupArray()
+```
+
+**Note:** If you need to clean up resources after the benchmark run, you should use the corresponding `[GlobalCleanup]` attribute. It's going to be executed only once, after all benchmark iterations.
+
+### IterationSetup
+
+If your benchmark requires a clean state for every invocation, you need to use the `[IterationSetup]` attribute. Unfortunately, just using the `[IterationSetup]` attribute is not enough to get stable results. You also need to make sure that the benchmark itself performs enough of computations for a single invocation to run longer than 100ms. The closer to 250ms the more stable results you are going to get.
```

Comment:
I feel like this needs a glossary at the top: what is an iteration versus an invocation? etc. Terms to define are at least:

- iteration
- invocation
- duration (total time reported)
- duration of single invocation (actual metric we track)
...

Reply:
Good point, I extended the BDN summary with a link to [this BDN doc](https://benchmarkdotnet.org/articles/guides/how-it-works.html) that explains what these things are using words and pseudocode.